clippy_lints/src/methods/map_flatten.rs

```diff
@@ -1,83 +1,119 @@
+    let map_closure_ty = cx.typeck_results().expr_ty(map_arg);
+    match map_closure_ty.kind() {
+        ty::Closure(_, _) | ty::FnDef(_, _) | ty::FnPtr(_) => {
+            let map_closure_sig = match map_closure_ty.kind() {
+                ty::Closure(_, substs) => substs.as_closure().sig(),
+                _ => map_closure_ty.fn_sig(cx.tcx),
+            };
+            let map_closure_return_ty = cx.tcx.erase_late_bound_regions(map_closure_sig.output());
+            is_type_diagnostic_item(cx, map_closure_return_ty, sym::Option)
         },
-        _ => {
-            return;
-        },
-    };
+        _ => false,
+    }
+}
 
-    let func_snippet = snippet(cx, map_arg.span, "..");
-    let hint = format!(".and_then({})", func_snippet);
-    let lint_info = format!("called `map(..).flatten()` on an `{}`", caller_type);
-    span_lint_and_sugg(
-        cx,
-        MAP_FLATTEN,
-        expr.span.with_lo(recv.span.hi()),
-        &lint_info,
-        "try using `and_then` instead",
-        hint,
-        Applicability::MachineApplicable,
-    );
+/// Split code suggestion into upper and bottom parts if its number of lines exceeds
+/// `MAX_SUGGESTION_HIGHLIGHT_LINES`, otherwise this function will just call `span_lint_and_sugg`
+/// with a combined `help` message.
+///
+/// This is useful when a suggested code snippet has multiple line thus causing possible useful
+/// information being chopped off at the end.
+fn span_lint_and_splitted_sugg(
+    cx: &LateContext<'_>,
+    sp: Span,
+    msg: &str,
+    help: &[&str; 2],
+    sugg: String,
+    applicability: Applicability,
+) {
+    let sugg_lines_count = sugg.lines().count();
+    if sugg_lines_count > MAX_SUGGESTION_HIGHLIGHT_LINES {
+        let sm = cx.sess().source_map();
+        if let (Ok(src_line_upper), Ok(src_line_bottom)) = (sm.lookup_line(sp.lo()), sm.lookup_line(sp.hi())) {
+            let split_idx = MAX_SUGGESTION_HIGHLIGHT_LINES / 2;
+            let span_upper = sp.with_hi(src_line_upper.sf.lines[src_line_upper.line + split_idx]);
+            // span over the rest of the last line of `span_upper`
+            let span_upper = sm.span_until_char(span_upper, '\n');
+            let span_bottom = sp.with_lo(src_line_bottom.sf.lines[src_line_bottom.line - split_idx]);
+
+            let sugg_lines_vec = sugg.lines().collect::<Vec<&str>>();
+            let sugg_upper = sugg_lines_vec[..split_idx].join("\n");
+            let sugg_bottom = sugg_lines_vec[sugg_lines_count - split_idx..].join("\n");
+
+            span_lint_and_then(cx, MAP_FLATTEN, sp, msg, |diag| {
+                diag.span_suggestion(span_upper, help[0], sugg_upper, applicability);
+                diag.span_suggestion(span_bottom, help[1], sugg_bottom, applicability);
+            });
+            return;
+        }
+    }
+    let combined_help = help.join(", ");
+    span_lint_and_sugg(cx, MAP_FLATTEN, sp, msg, &combined_help, sugg, applicability);
```

Comment:
This help message can get quite log when the two messages are joined together. I think it might be better to add another message just for this case or use `span_lint_and_then` to call [`span_suggestion_with_style`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/struct.DiagnosticBuilder.html#method.span_suggestion_with_style) to set the suggestion display style to [`ShowAlways`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/enum.SuggestionStyle.html#variant.ShowAlways). Then the suggestion would always be shown in a help "window" like this:

```rs
= help: <first_msg>, <second_msg>
   |
LL + vec![5_i8; 6].into_iter().filter_map(|x| x.checked_add(1)).collect();
   | 
```